lgc/disassembler/Disassembler.cpp

```diff
@@ -64,6 +65,62 @@ const char *getPalMetadataRegName(unsigned regNumber);
 
 namespace {
 
+struct InstOp {
+  Optional<int64_t> imm;
+  Optional<unsigned> sReg;
+};
+
+struct InstOrDirective {
+  MCDisassembler::DecodeStatus status = MCDisassembler::Fail;
+  uint64_t offset;
+  ArrayRef<uint8_t> bytes;
+  MCInst mcInst;
+
+  StringRef mnemonic;
+  StringRef comment;
+  InstOp op0;
+  InstOp op1;
+  InstOp op2;
+
+  const MCExpr *valueDirectiveExpr = nullptr;
+
+  uint64_t getEndOffset() const { return offset + bytes.size(); }
+};
+
+class SymbolPool {
+public:
+  SymbolPool(StringSaver &strings, MCContext &context) : m_strings(strings), m_context(context) {}
+
+  MCSymbol *getOrCreateSymbol(uint64_t offset, Twine name = {}, unsigned type = ELF::STT_NOTYPE) {
+    MCSymbol *&sym = m_symbolMap[std::make_pair(offset, type)];
+    if (!sym) {
+      StringRef savedName = m_strings.save(!name.isTriviallyEmpty() ? name : "_L" + Twine::utohexstr(offset));
```

Comment:
Can we remove the negation and make the condition more straightforward? Or even factor out name generation into a separate helper function.